py3status/modules/air_quality.py

```diff
@@ -0,0 +1,170 @@
+
+Color options:
+    color_<category>: display color
+        eg color_hazardous = '#7E0023'
+
+Example:
+```
+air_quality {
+    auth_token = 'demo'
+    location = 'Shanghai'
+    format = 'Shanghai: {aqi} {category}'
+}
+```
+@author beetleman
+@license BSD
+"""
+
+try:
+    # python 3
+    from urllib.parse import urlencode
+except ImportError:
+    # python 2
+    from urllib import urlencode
+
+AQI = (
+    (0, 50, 'good'),
+    (51, 100, 'moderate'),
+    (101, 150, 'unhealthy_sensitive'),
+    (151, 200, 'unhealthy'),
+    (201, 300, 'very_unhealthy'),
+    (301, int('inf'), 'hazardous')
+)
+CATEGORY = {
+    'good': 'Good',
+    'hazardous': 'Hazardous',
+    'moderate': 'Moderate',
+    'unhealthy': 'Unhealthy',
+    'unhealthy_sensitive': 'Unhealthy for Sensitive Groups',
+    'unknown': 'Unknown',
+    'very_unhealthy': 'Very Unhealthy',
+}
+COLOR = {
+    'good': '#009966',
+    'hazardous': '#7E0023',
+    'moderate': '#FFDE33',
+    'unhealthy': '#CC0033',
+    'unhealthy_sensitive': '#FF9933',
+    'unknown': '#ffffff',
+    'very_unhealthy': '#660099',
+}
+STRING_UNKNOWN = 'Unknown'
+URL = 'http://api.waqi.info'
+
+
+def _get_in(coll, path, default=None):
+    first = path[0]
+    rest = path[1:len(path)]
+    if len(path) == 1:
+        return coll.get(first, default)
+    return _get_in(coll.get(first, default), rest, default)
+
+
+class Py3status:
+    auth_token = 'demo'
+    cache_timeout = 3600
+    format = 'AQI: {aqi} {category}'
+    location = 'Shanghai'
+
+    def _update_aqi(self):
+        url_path = 'feed/{}'.format(self.location)
+        url_token = {'token': self.auth_token}
+        url_full = '{url_base}/{url_path}/?{url_query}'.format(
+            url_base=URL,
+            url_path=url_path,
+            url_query=urlencode(url_token)
+        )
+        try:
+            self._api_data = self.py3.request(url_full).json()
+        except (self.py3.RequestException):
+            self._api_data = None
+
+    def _update_key(self):
+        if self._api_data.get('status') == 'ok':
+            aqi = _get_in(self._api_data, ['data', 'aqi'], -1)
+            if type(aqi) is not int:
+                self._key = 'unknown'
+                return
+            for start, end, key in AQI:
+                if aqi >= start and aqi <= end:
+                    self._key = key
+                    return
+        self._key = 'unknown'
+
+    def _get_full_text(self):
+        aqi = _get_in(self._api_data, ['data', 'aqi'])
+        iaqi = _get_in(self._api_data, ['data', 'iaqi'], {})
+        try:
+            category = getattr(self, 'CATEGORY_{}'.format(self._key).lower())
+        except:
+            category = CATEGORY.get(self._key)
+        if not category:
+            category = STRING_UNKNOWN
+        return self.format.format(
+            location=self.location,
+            category=category,
+            aqi=aqi,
+            **{k: _v.get('v') for k, _v in iaqi.items()}
```

Comment:
something like

        return self.py3.safe_format(self.format,
                               dict(location=self.location,
                                    category=category,
                                    aqi=aqi,
            **{k: v.get('v') for k, v in iaqi.items()})
        )

but maybe do the `{k: v.get('v') for k, v in iaqi.items()}` bit before this function to be more readable